src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,665 @@
+            }
+            return HASH_NOT_FOUND;
+        }
+    }
+
+    @GenerateUncached
+    @ImportStatic(CompactHashStore.class)
+    abstract static class SetKvAtNode extends RubyBaseNode {
+
+        public abstract boolean execute(RubyHash hash, CompactHashStore store, int kvPos, int keyHash, Object frozenKey,
+                Object value);
+
+        // key already exist, very fast because there is no need to touch the index
+        @Specialization(guards = "kvPos != KEY_NOT_FOUND")
+        boolean keyAlreadyExistsWithDifferentValue(
+                RubyHash hash, CompactHashStore store, int kvPos, int keyHash, Object frozenKey, Object value) {
+            store.kvStore[kvPos + 1] = value;
+            return false;
+        }
+
+        // setting the key is a relatively expensive insertion
+        @Specialization(guards = "kvPos == KEY_NOT_FOUND")
+        static boolean keyDoesntExist(
+                RubyHash hash, CompactHashStore store, int kvPos, int keyHash, Object frozenKey, Object value,
+                @Cached @Exclusive InlinedConditionProfile kvResizingIsNeeded,
+                @Cached @Exclusive InlinedConditionProfile indexResizingIsNotNeeded,
+                @Cached @Exclusive InlinedLoopConditionProfile indexSlotUnavailable,
+                @Cached @Exclusive InlinedLoopConditionProfile idxSlotUnavailable,
+                @Cached @Exclusive InlinedLoopConditionProfile tillIndexArrayEnd,
+                @Bind("$node") Node node) {
+            resizeIndexIfNeeded(store, hash.size,
+                    indexResizingIsNotNeeded, idxSlotUnavailable, tillIndexArrayEnd, node);
+            resizeKvStoreIfNeeded(store, kvResizingIsNeeded, node);
+
+            int pos = store.kvStoreInsertionPos;
+            insertIntoKv(store, frozenKey, value);
+            insertIntoIndex(keyHash, pos + 1, store.index,
+                    indexSlotUnavailable, node);
+
+            hash.size++;
+            return true;
+        }
+
+        private static void insertIntoIndex(int keyHash, int kvPos, int[] index,
+                InlinedLoopConditionProfile unavailableSlot, Node node) {
+            int pos = getIndexPosFromHash(keyHash, index);
+
+            boolean slotFilled = index[pos + 1] > INDEX_SLOT_UNUSED;
+            while (unavailableSlot.profile(node, slotFilled)) {
+                pos = incrementIndexPos(pos, index.length);
+                slotFilled = index[pos + 1] > INDEX_SLOT_UNUSED;
+            }
+
+            index[pos] = keyHash;
+            index[pos + 1] = kvPos;
+        }
+
+        private static void insertIntoKv(CompactHashStore store, Object key, Object value) {
+            store.kvStore[store.kvStoreInsertionPos] = key;
+            store.kvStore[store.kvStoreInsertionPos + 1] = value;
+            store.kvStoreInsertionPos += 2;
+        }
+
+        @TruffleBoundary
+        private static void resizeIndexIfNeeded(CompactHashStore store, int size,
```

Comment:
The heavy logic with the loop below, that's good to be behind a TruffleBoundary, there is almost no value to partial evaluate that, and it would make warmup slower.
But the early check `if (indexResizingIsNotNeeded.profile(node, size < store.numSlotsForIndexRebuild)) {` that must be moved to each caller, we don't want to make a boundary call if it's not needed.